org.elasticsearch.xpack.transform.integration.TestFeatureResetIT

### CI Link

https://gradle-enterprise.elastic.co/s/wmkqnkogroefy

### Repro line

./gradlew ':x-pack:plugin:transform:qa:multi-node-tests:javaRestTest' --tests "org.elasticsearch.xpack.transform.integration.TestFeatureResetIT.testTransformFeatureReset" -Dtests.seed=A5D75D753CF49FA6 -Dtests.locale=cs -Dtests.timezone=Asia/Famagusta -Druntime.java=21	

### Does it reproduce?

No

### Applicable branches

main

### Failure history

_No response_

### Failure excerpt

It does seem that a transforms index is being created again after the feature reset. Here are the cluster logs from this failure:
[transform.zip](https://github.com/elastic/elasticsearch/files/12857364/transform.zip)

```
org.elasticsearch.xpack.transform.integration.TestFeatureResetIT > testTransformFeatureReset FAILED |  
-- | --
  | java.lang.AssertionError: |  
  | Expected: is a map with size <0> |  
  | but: map size was <1> |  
  | at __randomizedtesting.SeedInfo.seed([A5D75D753CF49FA6:B8C2ACD965E57A12]:0) |  
  | at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18) |  
  | at org.junit.Assert.assertThat(Assert.java:956) |  
  | at org.junit.Assert.assertThat(Assert.java:923) |  
  | at org.elasticsearch.xpack.transform.integration.TestFeatureResetIT.testTransformFeatureReset(TestFeatureResetIT.java:117)
```

> Yes, I was thinking of a notification too. Unfortunately, the test assertion didn't tell us which transform-* index existed where it should not. I'll fix this.

So I now have confirmation that it is indeed `.transform-notifications-000002` index being created:

```
org.elasticsearch.xpack.transform.integration.TestFeatureResetIT > testTransformFeatureReset FAILED
    java.lang.AssertionError: Indices were: {.transform-notifications-000002={settings={index={routing={allocation={include={_tier_preference=data_content}}}, hidden=true, number_of_shards=1, auto_expand_replicas=0-1, provided_name=.transform-notifications-000002, creation_date=1697002490163, number_of_replicas=1, uuid=6t9YgmmmQ_2xvGuhTaLWLA, version={created=8500003}}}, mappings={_meta={managed_index_mappings_version=1, version=8.11.0}, dynamic=false, properties={level={type=keyword}, transform_id={type=keyword}, node_name={type=keyword}, message={type=text, fields={raw={ignore_above=1024, type=keyword}}}, timestamp={type=date}}}, aliases={.transform-notifications-read={is_hidden=true}}}}
    Expected: is a map with size <0>
         but: map size was <1>
        at __randomizedtesting.SeedInfo.seed([61A84AAC7843C044:7CBDBB00215225F0]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.elasticsearch.xpack.transform.integration.TestFeatureResetIT.testTransformFeatureReset(TestFeatureResetIT.java:118)
```

Surprisingly, it didn't have any documents (audit messages) in it. Otherwise, the previous assertion (that searches for index contents) would fail while it didn't.
So I'm looking for a place where an **empty** notifications index is created.

--- a/docs/changelog/104870.yaml
+++ b/docs/changelog/104870.yaml
@@ -1,0 +1,7 @@
+pr: 104870
+summary: Make `_reset` action stop transforms without force first
+area: Transform
+type: bug
+issues:
+ - 100596
+ - 104825

--- a/x-pack/plugin/transform/qa/multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/transform/integration/TestFeatureResetIT.java
+++ b/x-pack/plugin/transform/qa/multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/transform/integration/TestFeatureResetIT.java
@@ -90,6 +90,7 @@ public class TestFeatureResetIT extends TransformRestTestCase {
 
         putTransform(continuousTransformId, Strings.toString(config), RequestOptions.DEFAULT);
 
+        // Sleep for a few seconds so that we cover transform being stopped at various stages.
         Thread.sleep(randomLongBetween(0, 5_000));
 
         startTransform(continuousTransformId, RequestOptions.DEFAULT);

--- a/x-pack/plugin/transform/qa/multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/transform/integration/TestFeatureResetIT.java
+++ b/x-pack/plugin/transform/qa/multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/transform/integration/TestFeatureResetIT.java
@@ -57,7 +57,6 @@ public class TestFeatureResetIT extends TransformRestTestCase {
         cleanUp();
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/104825")
     @SuppressWarnings("unchecked")
     public void testTransformFeatureReset() throws Exception {
         String indexName = "basic-crud-reviews";
@@ -90,6 +89,9 @@ public class TestFeatureResetIT extends TransformRestTestCase {
             .build();
 
         putTransform(continuousTransformId, Strings.toString(config), RequestOptions.DEFAULT);
+
+        Thread.sleep(randomLongBetween(0, 5_000));
+
         startTransform(continuousTransformId, RequestOptions.DEFAULT);
 
         assertOK(client().performRequest(new Request(HttpPost.METHOD_NAME, "/_features/_reset")));

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/Transform.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/Transform.java
@@ -404,7 +404,7 @@ public class Transform extends Plugin implements SystemIndexPlugin, PersistentTa
             SystemIndexPlugin.super.cleanUpFeature(clusterService, client, unsetResetModeListener);
         }, unsetResetModeListener::onFailure);
 
-        ActionListener<StopTransformAction.Response> afterStoppingTransforms = ActionListener.wrap(stopTransformsResponse -> {
+        ActionListener<StopTransformAction.Response> afterForceStoppingTransforms = ActionListener.wrap(stopTransformsResponse -> {
             if (stopTransformsResponse.isAcknowledged()
                 && stopTransformsResponse.getTaskFailures().isEmpty()
                 && stopTransformsResponse.getNodeFailures().isEmpty()) {
@@ -439,12 +439,28 @@ public class Transform extends Plugin implements SystemIndexPlugin, PersistentTa
             }
         }, unsetResetModeListener::onFailure);
 
+        ActionListener<StopTransformAction.Response> afterStoppingTransforms = ActionListener.wrap(
+            afterForceStoppingTransforms::onResponse,
+            e -> {
+                logger.warn("Exception while trying to stop the transforms, will try again with force=true", e);
+                StopTransformAction.Request forceStopTransformsRequest = new StopTransformAction.Request(
+                    Metadata.ALL,
+                    true,
+                    true,
+                    null,
+                    true,
+                    false
+                );
+                client.execute(StopTransformAction.INSTANCE, forceStopTransformsRequest, afterForceStoppingTransforms);
+            }
+        );
+
         ActionListener<AcknowledgedResponse> afterResetModeSet = ActionListener.wrap(response -> {
             StopTransformAction.Request stopTransformsRequest = new StopTransformAction.Request(
                 Metadata.ALL,
                 true,
-                true,
-                null,
+                false,
+                TimeValue.timeValueSeconds(10),
                 true,
                 false
             );

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/Transform.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/Transform.java
@@ -442,10 +442,11 @@ public class Transform extends Plugin implements SystemIndexPlugin, PersistentTa
         ActionListener<StopTransformAction.Response> afterStoppingTransforms = ActionListener.wrap(
             afterForceStoppingTransforms::onResponse,
             e -> {
-                logger.warn("Exception while trying to stop the transforms, will try again with force=true", e);
+                logger.info("Error while trying to stop the transforms, will try again with force=true", e);
                 StopTransformAction.Request forceStopTransformsRequest = new StopTransformAction.Request(
                     Metadata.ALL,
                     true,
+                    // Set force=true to make sure all the transforms persistent tasks are stopped.
                     true,
                     null,
                     true,
@@ -459,7 +460,9 @@ public class Transform extends Plugin implements SystemIndexPlugin, PersistentTa
             StopTransformAction.Request stopTransformsRequest = new StopTransformAction.Request(
                 Metadata.ALL,
                 true,
+                // Set force=false in order to let transforms finish gracefully.
                 false,
+                // Do not give it too much time. If there is a problem, there will be another try with force=true.
                 TimeValue.timeValueSeconds(10),
                 true,
                 false

